ztools/iscsi/iscsi.go

```diff
@@ -0,0 +1,332 @@
+package iscsi
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strings"
+	//	"net"
+)
+
+const (
+	// Initiator opcodes
+	NOP_OUT               byte = 0
+	ISCSI_COMMAND         byte = 1
+	ISCSI_TASK_MANAGEMENT byte = 2
+	LOGIN_REQUEST         byte = 3
+	TEXT_REQUEST          byte = 4
+	SCSI_DATA_OUT         byte = 5
+	LOGOUT_REQUEST        byte = 6
+	SNACK_REQUEST         byte = 0x10
+
+	// Target opcodes
+	NOP_IN               byte = 0x20
+	SCSI_RESPONSE        byte = 0x21
+	SCSI_TASK_MANAGEMENT byte = 0x22
+	LOGIN_RESPONSE       byte = 0x23
+	TEXT_RESPONSE        byte = 0x24
+	SCSI_DATA_IN         byte = 0x25
+	LOGOUT_RESPONSE      byte = 0x26
+	READY_TO_TRANSFER    byte = 0x31
+	ASYNCHRONOUS_MESSAGE byte = 0x32
+	REJECT               byte = 0x3f
+
+	// Special Bits, which need to be added to certain opcodes or fields
+	REQUEST_PDU byte = 0x40 // added to opcode denotes request
+	FINAL       byte = 0x80 // added to first field denotes final (or only) PDU
+	TRANSIT     byte = 0x80 // added to first field denotes readiness to transit to next stage
+	CONTINUE    byte = 0x40 // added to first field denotes that request is incomplete
+
+	NSG_SECURITY byte = 0x00
+	NSG_LOGIN    byte = 0x01
+	NSG_FULL     byte = 0x03
+
+	CSG_SECURITY byte = 0x00
+	CSG_LOGIN    byte = 0x04
+	CSG_FULL     byte = 0xc
+
+	ISID_OUI      = 0x00
+	ISID_EN       = 0x40
+	ISID_RANDOM   = 0x80
+	ISID_RESERVED = 0xc0
+
+	// AHSTypes
+	EXTENDED_CBD                            byte = 0x01
+	EXPECTED_BIDIRECTIONAL_READ_DATA_LENGTH byte = 0x02
+
+	VERSION_MAX byte = 0
+	VERSION_MIN byte = 0
+)
+
+var (
+	// Cisco OUI as per IEEE
+	ISID_CISCO_OUI [6]byte = [6]byte{0, 0x02, 0x3d, 0, 0, 0}
+	TSIH_LOGIN     [2]byte = [2]byte{0, 0}
+)
+
+type Header interface {
+	// empty for now
+}
+
+type BasicHeader struct {
+	Opcode            byte
+	Flags             byte
+	VersionMax        byte
+	VersionMin        byte
+	TotalAHSLength    byte
+	DataSegmentLength [3]byte
+}
+
+type TextParameter struct {
+	Key   string
+	Value string
+}
+
+func (t *TextParameter) GetLength() {
```

Comment:
Header had to be an "easy to implement" interface which allowed me to keep pdu.Marshal and pdu.Unmarshal in a generic way but supports all the different header types. Each header type "implements" this interface since they are only made up of bytes or byte arrays which binary.Read and binary.Write know how to (de)serialize. It is a bit weird to do it this way but it works just fine.